### PR TITLE
Migrate environment variables for S3

### PR DESCRIPTION
**Pull request by ahlag**
Signed-off-by: changyonglik <theeahlag@gmail.com>

## Related Issues/PRs

https://github.com/mlflow/mlflow/issues/6379

## What changes are proposed in this pull request?
Migrate envs to `mlflow/environment_variables.py`
(Please fill in changes proposed in this fix)

## How is this patch tested?
TBD

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/environment_variables.py [hunk 1]
```diff
@@ -66,6 +66,21 @@ def __repr__(self):
 #: (default: ``False``)
```
Comment (dbczumar) [inline mlflow/environment_variables.py:66]: Nit, while we're here, can we change the documented default to ``""``?

File: mlflow/environment_variables.py [hunk 2]
```diff
@@ -66,6 +66,21 @@ def __repr__(self):
...
+#: (default: ``False``)
+MLFLOW_KERBEROS_TICKET_CACHE = _EnvironmentVariable("MLFLOW_KERBEROS_TICKET_CACHE", str, "")
+
+#: Specify whether to get kerberos user.
+#: (default: ``False``)
+MLFLOW_KERBEROS_USER = _EnvironmentVariable("MLFLOW_KERBEROS_USER", str, "")
+
+#: Specify whether to get extra pyarrow configurations.
+#: (default: ``False``)
+MLFLOW_PYARROW_EXTRA_CONF = _EnvironmentVariable("MLFLOW_PYARROW_EXTRA_CONF", str, "")
```
Comment (dbczumar) [inline mlflow/environment_variables.py:83]:
```suggestion
#: Specify whether to get extra S3 upload arguments.
#: (default: ``None``)
MLFLOW_S3_UPLOAD_EXTRA_ARGS = _EnvironmentVariable("MLFLOW_S3_UPLOAD_EXTRA_ARGS", str, None)

#: Specify whether to get kerberos extra cache.
#: (default: ``None``)
MLFLOW_KERBEROS_TICKET_CACHE = _EnvironmentVariable("MLFLOW_KERBEROS_TICKET_CACHE", str, None)

#: Specify whether to get kerberos user.
#: (default: ``None``)
MLFLOW_KERBEROS_USER = _EnvironmentVariable("MLFLOW_KERBEROS_USER", str, None)

#: Specify whether to get extra pyarrow configurations.
#: (default: ``None``)
MLFLOW_PYARROW_EXTRA_CONF = _EnvironmentVariable("MLFLOW_PYARROW_EXTRA_CONF", str, None)
```

File: mlflow/store/artifact/s3_artifact_repo.py [hunk 3]
```diff
@@ -76,7 +77,7 @@ def parse_s3_uri(uri):
     def get_s3_file_upload_extra_args():
         import json
 
-        s3_file_upload_extra_args = os.environ.get("MLFLOW_S3_UPLOAD_EXTRA_ARGS")
+        s3_file_upload_extra_args = MLFLOW_S3_UPLOAD_EXTRA_ARGS.get()
```
Comment (dbczumar) [inline mlflow/store/artifact/s3_artifact_repo.py:80]: Looks like we might be missing a few other environment variables: https://github.com/mlflow/mlflow/blob/6c1882d5b4f50bdb0c3694a4d12dddc5e16af4ff/mlflow/store/artifact/s3_artifact_repo.py#L87-L101

File: mlflow/projects/backend/local.py [hunk 4]
```diff
@@ -360,9 +360,9 @@ def _get_hdfs_artifact_cmd_and_envs(artifact_repo):
     # pylint: disable=unused-argument
     cmds = []
     envs = {
-        "MLFLOW_KERBEROS_TICKET_CACHE": os.environ.get("MLFLOW_KERBEROS_TICKET_CACHE"),
-        "MLFLOW_KERBEROS_USER": os.environ.get("MLFLOW_KERBEROS_USER"),
-        "MLFLOW_PYARROW_EXTRA_CONF": os.environ.get("MLFLOW_PYARROW_EXTRA_CONF"),
+        "MLFLOW_KERBEROS_TICKET_CACHE": MLFLOW_KERBEROS_TICKET_CACHE.get(),
```
Comment (dbczumar) [inline mlflow/projects/backend/local.py:363]: Looks like we need to import these environment var references from `mlflow.environment_variables`.

**Comment by ahlag** [comment]
Hi @dbczumar 
Thank you for the feedback! The PR is ready for review.

File: mlflow/environment_variables.py [hunk 5]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
 MLFLOW_GCS_DEFAULT_TIMEOUT = _EnvironmentVariable("MLFLOW_GCS_DEFAULT_TIMEOUT", int, None)
 
 #: Specify whether to disable model logging and loading via mlflowdbfs.
-#: (default: ``False``)
+#: (default: ``)
 _DISABLE_MLFLOWDBFS = _EnvironmentVariable("DISABLE_MLFLOWDBFS", str, "")
```
Comment (dbczumar) [inline mlflow/environment_variables.py:67]: Sorry, we recently made a change to the default value here; it is now `None`.

```suggestion
#: (default: `None`)
_DISABLE_MLFLOWDBFS = _EnvironmentVariable("DISABLE_MLFLOWDBFS", str, None)
```

File: mlflow/store/artifact/s3_artifact_repo.py [hunk 6]
```diff
@@ -11,7 +11,13 @@
 from mlflow.exceptions import MlflowException
 from mlflow.store.artifact.artifact_repo import ArtifactRepository
 from mlflow.utils.file_utils import relative_path_to_artifact_path
+from mlflow.environment_variables import MLFLOW_S3_UPLOAD_EXTRA_ARGS
```
Comment (dbczumar) [inline mlflow/store/artifact/s3_artifact_repo.py:14]: As indicated in the lint error, I think we should remove this duplicate import

**Review by dbczumar — APPROVED** [review]
LGTM once two small comments are addressed and the PR is merged / rebased with `master`. Thank you for this excellent contribution, @ahlag !

File: mlflow/environment_variables.py [hunk 7]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
...
 
+#: Specify whether to get extra S3 upload arguments.
```
Comment (dbczumar) [inline mlflow/environment_variables.py:69]:
```suggestion
#: Specify extra arguments for S3 artifact uploads.
```

File: mlflow/environment_variables.py [hunk 8]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
...
+#: (default: ``None``)
+MLFLOW_S3_UPLOAD_EXTRA_ARGS = _EnvironmentVariable("MLFLOW_S3_UPLOAD_EXTRA_ARGS", str, None)
+
+#: Specify whether to get kerberos extra cache.
```
Comment (dbczumar) [inline mlflow/environment_variables.py:73]:
```suggestion
#: Specifies the location of a Kerberos ticket cache to use for S3 artifact operations.
```

File: mlflow/environment_variables.py [hunk 9]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
...
+#: (default: ``None``)
+MLFLOW_KERBEROS_TICKET_CACHE = _EnvironmentVariable("MLFLOW_KERBEROS_TICKET_CACHE", str, None)
+
+#: Specify whether to get kerberos user.
```
Comment (dbczumar) [inline mlflow/environment_variables.py:77]:
```suggestion
#: Specifies a Kerberos user for S3 artifact operations.
```

File: mlflow/projects/backend/local.py [hunk 10]
```diff
@@ -360,9 +365,9 @@ def _get_hdfs_artifact_cmd_and_envs(artifact_repo):
...
```
Comment (dbczumar) [inline mlflow/projects/backend/local.py:368]: Can we also update references to these environment variables in https://github.com/mlflow/mlflow/blob/91899ac0f907da34ce0e8b952dc69db6a2f7da23/mlflow/store/artifact/hdfs_artifact_repo.py#L172-L174?

File: mlflow/environment_variables.py [hunk 11]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
...
+#: (default: ``None``)
+MLFLOW_PYARROW_EXTRA_CONF = _EnvironmentVariable("MLFLOW_PYARROW_EXTRA_CONF", str, None)
+
+#: Specify whether to get mlflow s3 endpoint url.
+#: (default: ``None``)
+MLFLOW_S3_ENDPOINT_URL = _EnvironmentVariable("MLFLOW_S3_ENDPOINT_URL", str, None)
+
+#: Specify whether to get mlflow s3 ignore tls configurations.
+#: (default: ``None``)
+MLFLOW_S3_IGNORE_TLS = _EnvironmentVariable("MLFLOW_S3_IGNORE_TLS", str, None)
```
Comment (dbczumar) [inline mlflow/environment_variables.py:91]: Can we move these below `MLFLOW_S3_UPLOAD_EXTRA_ARGS`?

File: mlflow/environment_variables.py [hunk 12]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
...
+
+#: Specify whether to get kerberos user.
+#: (default: ``None``)
+MLFLOW_KERBEROS_USER = _EnvironmentVariable("MLFLOW_KERBEROS_USER", str, None)
+
+#: Specify whether to get extra pyarrow configurations.
+#: (default: ``None``)
+MLFLOW_PYARROW_EXTRA_CONF = _EnvironmentVariable("MLFLOW_PYARROW_EXTRA_CONF", str, None)
+
+#: Specify whether to get mlflow s3 endpoint url.
```
Comment (dbczumar) [inline mlflow/environment_variables.py:85]:
```suggestion
#: Specifies the S3 endpoint URL to use for S3 artifact operations.
```

File: mlflow/environment_variables.py [hunk 13]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
...
+#: (default: ``None``)
+MLFLOW_S3_ENDPOINT_URL = _EnvironmentVariable("MLFLOW_S3_ENDPOINT_URL", str, None)
+
+#: Specify whether to get mlflow s3 ignore tls configurations.
+#: (default: ``None``)
```
Comment (dbczumar) [inline mlflow/environment_variables.py:90]:
```suggestion
#: (default: ``False``)
```

File: mlflow/environment_variables.py (hunk 11 above)
Comment (dbczumar) [inline mlflow/environment_variables.py:91]:
```suggestion
MLFLOW_S3_IGNORE_TLS = _EnvironmentVariable("MLFLOW_S3_IGNORE_TLS", bool, False)
```

File: mlflow/environment_variables.py [hunk 14]
```diff
@@ -63,9 +63,32 @@ def __repr__(self):
...
+
+#: Specify whether to get extra pyarrow configurations.
+#: (default: ``None``)
+MLFLOW_PYARROW_EXTRA_CONF = _EnvironmentVariable("MLFLOW_PYARROW_EXTRA_CONF", str, None)
+
+#: Specify whether to get mlflow s3 endpoint url.
+#: (default: ``None``)
+MLFLOW_S3_ENDPOINT_URL = _EnvironmentVariable("MLFLOW_S3_ENDPOINT_URL", str, None)
+
+#: Specify whether to get mlflow s3 ignore tls configurations.
```
Comment (dbczumar) [inline mlflow/environment_variables.py:89]:
```suggestion
#: Specifies whether or not to skip TLS certificate verification for S3 artifact operations.
```

File: mlflow/environment_variables.py (hunk 11 above)
Comment (harupy) [reply mlflow/environment_variables.py:91]:
```suggestion
MLFLOW_S3_IGNORE_TLS = _BooleanEnvironmentVariable("MLFLOW_S3_IGNORE_TLS", False)
```

We can use `_BooleanEnvironmentVariable` here.

**Comment by ahlag** [comment]
@dbczumar  @harupy 
Done!

File: mlflow/environment_variables.py [hunk 15]
```diff
@@ -92,10 +92,33 @@ def get(self):
 #: (default: ``None``). If None, ``google.cloud.storage.constants._DEFAULT_TIMEOUT`` is used.
 MLFLOW_GCS_DEFAULT_TIMEOUT = _EnvironmentVariable("MLFLOW_GCS_DEFAULT_TIMEOUT", int, None)
 
-#: Specify whether to disable model logging and loading via mlflowdbfs.
-#: (default: ``None``)
+#: Specifies whether to disable model logging and loading via mlflowdbfs.
+#: (default: `None`)
```
Comment (harupy) [inline mlflow/environment_variables.py:96]:
```suggestion
#: (default: ``None``)
```

Comment (ahlag) [reply mlflow/environment_variables.py:96]: @harupy 
What is the difference between single and double quotes here?

File: mlflow/environment_variables.py [hunk 16]
```diff
@@ -92,10 +92,33 @@ def get(self):
...
+#: (default: ``None``)
+MLFLOW_PYARROW_EXTRA_CONF = _EnvironmentVariable("MLFLOW_PYARROW_EXTRA_CONF", str, None)
+
+#: Specifies whether to get mlflow s3 endpoint url.
+#: (default: ``None``)
+MLFLOW_S3_ENDPOINT_URL = _EnvironmentVariable("MLFLOW_S3_ENDPOINT_URL", str, None)
+
+#: Specifies whether to get mlflow s3 ignore tls configurations.
+#: (default: ``False``)
+MLFLOW_S3_IGNORE_TLS = _BooleanEnvironmentVariable("MLFLOW_S3_IGNORE_TLS", False)
```
Comment (harupy) [inline mlflow/environment_variables.py:121]: Looks like some of @dbczumar's suggestions haven't applied yet.

File: mlflow/environment_variables.py (hunk 15 above)
Comment (harupy) [reply mlflow/environment_variables.py]: image

File: mlflow/environment_variables.py (hunk 16 above)
Comment (ahlag) [reply mlflow/environment_variables.py:121]: @harupy 
Is it changing all to False?

Comment (harupy) [reply mlflow/environment_variables.py:121]: No, I unresolved the suggestions that haven't been applied yet.

File: mlflow/environment_variables.py (hunk 15 above)
Comment (ahlag) [reply mlflow/environment_variables.py]: thanks!

**Comment by ahlag** [comment]
@harupy 
Ok done!

**Review by harupy — APPROVED** [review]
LGTM!

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @ahlag !

**Comment by salinaaaaaa** [comment]
what helm chart should we use?

**Pull request merged**